### PR TITLE
SNOW-666110: optimize flatten union_by_name

### PR DESCRIPTION
**Pull request by sfc-gh-aalam**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes #SNOW-666110

2. Fill out the following pre-review checklist:

   - [x] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   Use `select_statement.set_operator` to ensure that we do not add multiple nesting levels as we have more `union_by_name` calls stacked

### CODE REVIEW

File: src/snowflake/snowpark/dataframe.py [hunk 1]
```diff
@@ -1650,18 +1650,18 @@ def _union_by_name_internal(
...
-        union_plan = UnionPlan(self._plan, right_child._plan, is_all)
         if self._session.sql_simplifier_enabled:
             df = self._with_plan(
-                SelectStatement(
-                    from_=SelectSnowflakePlan(
-                        snowflake_plan=union_plan, analyzer=self._session._analyzer
+                self._select_statement.set_operator(
+                    right_child._select_statement
+                    or SelectSnowflakePlan(
+                        other._plan, analyzer=self._session._analyzer
```
[the remark this answers is not in the document]
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/dataframe.py]: fixed it. thanks for catching that

**Comment by sfc-gh-yixie** [comment]
We found a test config problem so the simplifier tests are not run. Please hold off merging this PR until we correct the problem to run simplifier problem.

**Comment by sfc-gh-yixie** [comment]
> We found a test config problem so the simplifier tests are not run. Please hold off merging this PR until we correct the problem to run simplifier problem.

main branch has fixed the problem. We need to rerun all tests.

**Review by sfc-gh-yixie — APPROVED** [review]

**Review by sfc-gh-jdu — APPROVED** [review]

File: src/snowflake/snowpark/dataframe.py [hunk 2]
```diff
@@ -1650,18 +1650,18 @@ def _union_by_name_internal(
...
-                SelectStatement(
-                    from_=SelectSnowflakePlan(
-                        snowflake_plan=union_plan, analyzer=self._session._analyzer
+                self._select_statement.set_operator(
+                    right_child._select_statement
+                    or SelectSnowflakePlan(
+                        right_child._plan, analyzer=self._session._analyzer
                     ),
-                    analyzer=self._session._analyzer,
+                    operator=SET_UNION,
```
Comment (sfc-gh-yixie) [inline src/snowflake/snowpark/dataframe.py:1660]: This needs to be `SET_UNION_ALL` when `is_all` is True.

**Review by sfc-gh-yixie — CHANGES_REQUESTED** [review]

File: src/snowflake/snowpark/dataframe.py (hunk 2 above)
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/dataframe.py]: yep. I just realized this error too. Good to know that SQL simplifier tests are working as expected

**Review by sfc-gh-yixie — APPROVED** [review]

**Pull request merged**